javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -260,6 +261,27 @@ void setCurrentArray() {
         }
     }
 
+    BitMappedTrie<T> filter(Predicate<? super T> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+
+        final Object[] results = new Object[length()];
+        final MutableInt index = new MutableInt();
+        this.<T> visit((leaf, start, end) -> {
+            int resultIndex = index.val;
```

Comment:
I think we do not need `MutableInt`. Instead we could add a `visit` method that takes a param `index` and returns the updated `index`:

```java
@FunctionalInterface
interface LeafVisitor<T> {
    int visit(T leaf, int start, int end, int index);
}
```

The `visit` method then looks like this:

```java
    @SuppressWarnings("unchecked")
    <T2> int visit(LeafVisitor<T2[]> visitor) {
        int start = lastDigit(offset);
        int currentIndex = 0;
        for (int index = 0; index < length; ) {
            final T2[] leaf = (T2[]) getLeaf(index);
            final int end = Math.min(leaf.length, start + length - index);

            currentIndex = visitor.visit(leaf, start, end, currentIndex);

            index += end - start;
            start = 0;
        }
        return currentIndex;
    }
```

BEFORE:

```java
        ...
        final MutableInt index = new MutableInt();
        this.<T> visit((leaf, start, end) -> {
            int resultIndex = index.val;
            for (int i = start; i < end; i++) {
                final T value = leaf[i];
                if (predicate.test(value)) {
                    results[resultIndex++] = value;
                }
            }
            index.val = resultIndex;
        });

        return (length() == index.val) ? ...
```

AFTER:

```java
        ...
        final int resultLength = this.<T> visit((leaf, start, end, index) -> {
            for (int i = start; i < end; i++) {
                final T value = leaf[i];
                if (predicate.test(value)) {
                    results[index++] = value;
                }
            }
            return index;
        });

        return (length() == resultLength) ? ...
```

Note: Similar changes could be applied to all other methods that use `visit`.